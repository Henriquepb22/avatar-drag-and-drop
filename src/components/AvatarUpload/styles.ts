import styled, { css } from 'styled-components'

import * as Button from 'components/Button/styles'

type WrapperProps = {
    removeBorder: boolean
}

export const Wrapper = styled.div<WrapperProps>`
    ${({ theme, removeBorder }) => css`
        background: ${theme.colors.grayLight};
        border: 2px dashed ${theme.colors.border};
        border-radius: ${theme.radius.medium};
        padding: ${theme.spacings.large};
        width: 55rem;
        height: 18rem;
        position: relative;

        ${removeBorder &&
        css`
            border-color: transparent;
        `}
    `}
`

export const Content = styled.div`
    ${({ theme }) => css`
        display: grid;
        grid-template-columns: 11.4rem 1fr;
        grid-gap: ${theme.spacings.large};
        align-items: center;
    `}
`

export const Form = styled.form`
    ${({ theme }) => css`
        display: flex;
        flex-direction: column;
        margin-right: ${theme.spacings.large};

        > :not(:last-child) {
            margin-bottom: ${theme.spacings.large};
        }

        ${Button.Wrapper} {
            align-self: flex-end;
            width: 11rem;
        }
    `}
`

export const CloseButton = styled.button`
    ${({ theme }) => css`
        position: absolute;
        top: 2rem;
        right: 2rem;
        display: flex;
        align-items: center;
        justify-content: center;
        width: 1.4rem;
        height: 1.4rem;
        padding: ${theme.spacings.medium};
        border: 0;
        background: none;

        &:hover {
            cursor: pointer;
        }
    `}
`
