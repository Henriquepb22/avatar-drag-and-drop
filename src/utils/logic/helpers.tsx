const isImageFile = (file: File) => {
    const fileType = file.type
    const baseMimeType = fileType.split('/')[0]

    if (baseMimeType !== 'image') return false
    return true
}

export const getFileDataUrl = async (file: File): Promise<string> => {
    const reader = new FileReader()

    return new Promise((resolve, reject) => {
        const isImageType = isImageFile(file)

        if (isImageType) {
            reader.readAsDataURL(file)
            reader.onload = () => {
                if (reader.result) {
                    resolve(reader.result.toString())
                }
            }
        } else {
            reader.abort()
            reject(new Error('File is not an recognized image type'))
        }
    })
}

export const cropImage = async (
    imageSrc: string,
    zoomLevel: number
): Promise<string> => {
    const canvas = document.createElement('canvas')
    const context = canvas.getContext('2d')
    const image = new Image()

    return new Promise((resolve, reject) => {
        image.src = imageSrc

        image.onload = () => {
            if (context) {
                const needCompression =
                    Math.max(canvas.width, canvas.height) > 1000

                canvas.width = needCompression ? image.width / 2 : image.width
                canvas.height = needCompression
                    ? image.height / 2
                    : image.height

                context.drawImage(image, 0, 0, canvas.width, canvas.height)
                context.globalCompositeOperation = 'destination-in'
                context.beginPath()
                context.arc(
                    canvas.width * 0.5,
                    canvas.height * 0.5,
                    (Math.max(canvas.width, canvas.height) * 0.5) /
                        (zoomLevel * 2),
                    0,
                    2 * Math.PI
                )
                context.fill()
                context.globalCompositeOperation = 'source-over'

                resolve(canvas.toDataURL('image/png', 1))
            }
        }

        image.onerror = () => {
            reject(new Error('Something bad happened...'))
        }
    })
}
