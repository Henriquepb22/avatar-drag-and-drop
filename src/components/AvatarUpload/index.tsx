import { useState } from 'react'
import { cropImage, getFileDataUrl } from 'utils/logic/helpers'
import FileDropzone from 'components/FileDropzone'
import ErrorBox from 'components/ErrorBox'
import Button from 'components/Button'
import Slider from 'components/Slider'
import Image from 'components/Image'

import closeIcon from 'assets/img/icons/close.svg'
import * as S from './styles'

const AvatarUpload = () => {
    const [uploadedFile, setUploadedFile] = useState<string | null>(null)
    const [zoomLevel, setZoomLevel] = useState(5)
    const [error, setError] = useState(false)
    const [isSaved, setIsSaved] = useState(false)

    const handleFileUpload = async (file: File) => {
        setUploadedFile(null)
        setIsSaved(false)
        try {
            const imageData = await getFileDataUrl(file)
            setUploadedFile(imageData)
        } catch (error) {
            setError(true)
        }
    }

    const handleReset = () => {
        setUploadedFile(null)
        setError(false)
    }

    const handleSave = async () => {
        if (uploadedFile) {
            const croppedImage = await cropImage(uploadedFile, zoomLevel)
            setUploadedFile(croppedImage)
            setIsSaved(true)
        }
    }

    // Scaling image from 100% to 250%
    const handleCrop = (cropBy: number) => setZoomLevel((cropBy + 5) / 10)

    return (
        <S.Wrapper removeBorder={!!error || (!!uploadedFile && !isSaved)}>
            {isSaved && uploadedFile && (
                <Image
                    src={uploadedFile}
                    altText="you cropped logo"
                    zoomLevel={zoomLevel}
                />
            )}
            {error && <ErrorBox handleReset={handleReset} />}
            {((!error && !uploadedFile) || isSaved) && (
                <FileDropzone
                    isSaved={isSaved}
                    handleFileUpload={handleFileUpload}
                />
            )}
            {!!uploadedFile && !isSaved && (
                <S.Content>
                    <Image
                        src={uploadedFile}
                        altText="you uploaded logo"
                        zoomLevel={zoomLevel}
                    />
                    <S.Form>
                        <Slider onChange={handleCrop} />
                        <Button type="button" onClick={handleSave}>
                            Save
                        </Button>
                    </S.Form>
                </S.Content>
            )}
            {(error || !!uploadedFile) && !isSaved && (
                <S.CloseButton
                    type="button"
                    aria-label="reset"
                    onClick={handleReset}
                >
                    <img src={closeIcon} alt="close" />
                </S.CloseButton>
            )}
        </S.Wrapper>
    )
}

export default AvatarUpload
