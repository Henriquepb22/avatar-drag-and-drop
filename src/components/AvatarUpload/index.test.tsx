import { screen, waitFor } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import userEvent from '@testing-library/user-event'

import AvatarUpload from '.'

const imageFile = new File(['image file'], 'image.png', {
    type: 'image/png',
})

describe('<AvatarUpload />', () => {
    it('should render avatar upload correctly', () => {
        const { container } = renderWithTheme(<AvatarUpload />)

        expect(screen.getByLabelText(/organization logo/i)).toBeInTheDocument()
        expect(
            screen.getByLabelText(/drop the image here/i)
        ).toBeInTheDocument()
        expect(container.firstChild).toMatchSnapshot()
    })

    it('should show error when uploading files that are not images', async () => {
        const notImageFile = new File(['file test'], 'text.txt', {
            type: 'text/plain',
        })
        renderWithTheme(<AvatarUpload />)

        const fileInput = screen.getByLabelText(
            /drop the image here/i
        ) as HTMLInputElement

        await waitFor(() => {
            userEvent.upload(fileInput, notImageFile)
        })

        await waitFor(() => {
            expect(
                screen.getByText(/sorry, the upload failed/i)
            ).toBeInTheDocument()
            expect(
                screen.getByRole('button', { name: /try again/i })
            ).toBeInTheDocument()
        })
    })

    it("should reset entire process clicking on the 'X' icon", async () => {
        renderWithTheme(<AvatarUpload />)

        const fileInput = screen.getByLabelText(
            /drop the image here/i
        ) as HTMLInputElement

        await waitFor(() => {
            userEvent.upload(fileInput, imageFile)
        })

        await waitFor(() => {
            const resetButton = screen.getByRole('button', { name: /reset/i })
            userEvent.click(resetButton)
            expect(fileInput).not.toBeInTheDocument()
        })

        expect(
            screen.getByLabelText(/drop the image here/i)
        ).toBeInTheDocument()
    })
})
