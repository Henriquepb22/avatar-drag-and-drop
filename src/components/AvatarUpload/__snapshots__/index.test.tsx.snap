// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarUpload /> should render avatar upload correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  font-size: 1.4rem;
  line-height: 180%;
}

.c1:hover {
  cursor: pointer;
}

.c1:focus-within {
  outline: -webkit-focus-ring-color auto 1px;
}

.c2 {
  color: #495567;
  font-weight: 500;
}

.c2 > img {
  margin-right: 1.2rem;
}

.c3 {
  color: #677489;
}

.c4 {
  height: 0;
  width: 0;
  border: none;
  opacity: 0;
}

.c0 {
  background: #F2F5F8;
  border: 2px dashed #C7CDD3;
  border-radius: 0.8rem;
  padding: 3.2rem;
  width: 55rem;
  height: 18rem;
  position: relative;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="uploadInput"
  >
    <p
      class="c2"
    >
      <img
        alt="mountain and sun"
        src="painting.svg"
      />
      Organization Logo
    </p>
    <p
      class="c3"
    >
      Drop the image here or click to browse.
    </p>
    <input
      accept="image/*"
      class="c4"
      id="uploadInput"
      tabindex="0"
      type="file"
    />
  </label>
</div>
`;
